[JSC] Should use WeakPtr for Waiter's TicketData

hyjorc1 opened the pull request:
#### fce181433e3b5225ab4b92a9be2491415964291a
<pre>
[JSC] Should use WeakPtr for Waiter&apos;s TicketData
<a href="https://bugs.webkit.org/show_bug.cgi?id=276410">https://bugs.webkit.org/show_bug.cgi?id=276410</a>
<a href="https://rdar.apple.com/131442732">rdar://131442732</a>

Reviewed by Yusuke Suzuki.

Previously, we fixed the memory issue in DeferredWorkTimer::TicketData
by letting JSGlobalObject to strong reference TicketData&apos;s dependencies[1].
Later, we fixed the tracked TicketData with WeakHashSet in JSGlobalObject[2].
This patch fixed a similar issue by applying WeakPtr to Waiter&apos;s TicketData.
I&apos;ve confirmed that all DeferredWorkTimer::TicketData uses are guarded by
WeakPtr, WeakHashSet, Ref, or RefPtr.

[1] <a href="https://commits.webkit.org/279852@main">https://commits.webkit.org/279852@main</a>
[2] <a href="https://commits.webkit.org/280180@main">https://commits.webkit.org/280180@main</a>

* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::Waiter::scheduleWorkAndClearTicket):
(JSC::WaiterListManager::cancelAsyncWaiter):
(JSC::Waiter::dump const):
* Source/JavaScriptCore/runtime/WaiterListManager.h:

Canonical link: <a href="https://commits.webkit.org/280830@main">https://commits.webkit.org/280830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ef97f810b9ba6d824fa4293221e038666739801

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57775 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37103 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10251 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61397 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/8220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44739 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8408 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/8220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59805 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/34796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/49932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/27639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/31566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/7225 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7224 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/50867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/7495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63080 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/57017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1689 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/7562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1695 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/49943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/54147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/1432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78778 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8611 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/32932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/13066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34018 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/35102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33763 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -278,8 +278,8 @@ size_t WaiterListManager::totalWaiterCount()
-    auto ticket = waiter->ticket(listLocker);
-    waiter->vm()->deferredWorkTimer->cancelPendingWork(ticket);
+    RefPtr<DeferredWorkTimer::TicketData> ticket = waiter->ticket(listLocker);
+    waiter->vm()->deferredWorkTimer->cancelPendingWork(ticket.get());
```
Let's check whether ticket is nullptr or not.

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Should use WeakPtr for Waiter's TicketData
https://bugs.webkit.org/show_bug.cgi?id=276410
rdar://131442732

Reviewed by Yusuke Suzuki.

Previously, we fixed the memory issue in DeferredWorkTimer::TicketData
by letting JSGlobalObject to strong reference TicketData's dependencies[1].
Later, we fixed the tracked TicketData with WeakHashSet in JSGlobalObject[2].
This patch fixed a similar issue by applying WeakPtr to Waiter's TicketData.
I've confirmed that all DeferredWorkTimer::TicketData uses are guarded by
WeakPtr, WeakHashSet, Ref, or RefPtr.

[1] https://commits.webkit.org/279852@main
[2] https://commits.webkit.org/280180@main

* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::Waiter::scheduleWorkAndClearTicket):
(JSC::WaiterListManager::cancelAsyncWaiter):
(JSC::Waiter::dump const):
* Source/JavaScriptCore/runtime/WaiterListManager.h:

Canonical link: https://commits.webkit.org/280830@main
